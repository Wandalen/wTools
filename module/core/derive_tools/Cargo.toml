[package]
name = "derive_tools"
version = "0.27.0"
edition = "2021"
authors = [
  "Kostiantyn Wandalen <wandalen@obox.systems>",
  "Dmytro Kryvoruchko <dmytro.kr@obox.systems>",
]
license = "MIT"
readme = "Readme.md"
documentation = "https://docs.rs/derive_tools"
repository = "https://github.com/Wandalen/wTools/tree/master/module/core/derive_tools"
homepage = "https://github.com/Wandalen/wTools/tree/master/module/core/derive_tools"
description = """
A collection of derive macros designed to enhance STD.
"""
categories = [ "algorithms", "development-tools" ]
keywords = [ "fundamental", "general-purpose" ]


[lints]
workspace = true


[package.metadata.docs.rs]
features = [ "full" ]
all-features = false



[features]

default = [
  "enabled",

  "derive_variadic_from",
  "type_variadic_from",

  "derive_add_assign",
  "derive_add",
  "derive_as_mut",
  "derive_as_ref",
  "derive_constructor",
  "derive_deref",
  "derive_deref_mut",
  "derive_error",

  "derive_index",
  "derive_index_mut",
  # "derive_inner_from",
  "derive_into_iterator",
  # "derive_iterator",
  "derive_mul_assign",
  "derive_mul",
  "derive_not",
  "derive_sum",
  "derive_try_into",
  "derive_is_variant",
  "derive_unwrap",
  # "derive_convert_case",
  "derive_into",

  "derive_display",
  "derive_from_str",

  "derive_clone_dyn",
  # "derive_clone_dyn_use_std",

  "derive_strum",
  "strum_phf",

  "derive_from",
  "derive_inner_from",
  "derive_new",

  "derive_phantom",

  # "use_std",
]

full = [
  "enabled",

  "derive_variadic_from",
  "type_variadic_from",

  "derive_add_assign",
  "derive_add",
  "derive_as_mut",
  "derive_as_ref",
  "derive_constructor",
  "derive_deref",
  "derive_deref_mut",
  "derive_error",

  "derive_index",
  "derive_index_mut",
  # "derive_inner_from",
  "derive_into_iterator",
  # "derive_iterator",
  "derive_mul_assign",
  "derive_mul",
  "derive_not",
  "derive_sum",
  "derive_try_into",
  "derive_is_variant",
  "derive_unwrap",
  # "derive_convert_case",

  "derive_display",
  "derive_from_str",

  "derive_clone_dyn",
  # "derive_clone_dyn_use_std",

  "derive_strum",
  "strum_phf",

  "derive_from",
  "derive_inner_from",
  "derive_new",

  "derive_phantom",

  # "use_std",
]
no_std = []
use_alloc = [ "no_std" ]
enabled = [ "derive_tools_meta/enabled" ]

# nightly = [ "derive_more/nightly" ]

type_variadic_from = [ "variadic_from/enabled", "variadic_from/type_variadic_from" ]
derive_variadic_from = [ "type_variadic_from", "derive_tools_meta/derive_variadic_from", "variadic_from/derive_variadic_from" ]

# enable_track_caller = [ "derive_more", "derive_more/track-caller" ]

derive_add_assign = [ "derive_more", "derive_more/std", "derive_more/add_assign" ]
derive_add = [ "derive_more", "derive_more/std", "derive_more/add" ]
derive_as_mut = [ "derive_tools_meta/derive_as_mut" ]
derive_as_ref = [ "derive_tools_meta/derive_as_ref" ]
# derive_as_mut = [ "derive_more", "derive_more/as_mut" ]
# derive_as_ref = [ "derive_more", "derive_more/as_ref" ]
derive_constructor = [ "derive_more", "derive_more/std", "derive_more/constructor" ]
derive_deref = [ "derive_tools_meta/derive_deref" ]
derive_deref_mut = [ "derive_tools_meta/derive_deref_mut" ]
# derive_deref = [ "derive_more", "derive_more/deref" ]
# derive_deref_mut = [ "derive_more", "derive_more/deref_mut" ]
derive_error = [ "derive_more", "derive_more/std", "derive_more/error" ]
# derive_from = [ "derive_more", "derive_more/from" ]
# derive_from = [ "derive_tools_meta/derive_from" ]
# derive_reflect = [ "derive_tools_meta/derive_reflect" ]

# derive_index = [ "derive_more", "derive_more/std", "derive_more/index" ]
# derive_index_mut = [ "derive_more", "derive_more/std", "derive_more/index_mut" ]
# derive_inner_from = [ "derive_more", "derive_more/into" ]
derive_into_iterator = [ "derive_more", "derive_more/std", "derive_more/into_iterator" ]
# derive_iterator = [ "derive_more", "derive_more/iterator" ]
derive_into = [ "derive_more", "derive_more/into" ]
derive_mul_assign = [ "derive_more", "derive_more/std", "derive_more/mul_assign" ]
derive_mul = [ "derive_more", "derive_more/std", "derive_more/mul" ]
derive_sum = [ "derive_more", "derive_more/std", "derive_more/sum" ]
derive_try_into = [ "derive_more", "derive_more/std", "derive_more/try_into" ]
derive_is_variant = [ "derive_more", "derive_more/is_variant" ]
derive_unwrap = [ "derive_more", "derive_more/std", "derive_more/unwrap" ]
# derive_convert_case = [ "derive_more", "derive_more/convert_case" ]

derive_display = [ "parse-display" ]
derive_from_str = [ "parse-display", "parse-display/std", "parse-display/regex" ]
# derive_from_str = [ "parse-display", "parse-display/std", "parse-display/regex", "parse-display/once_cell" ]

derive_strum = [ "strum/std", "strum/derive", "strum/strum_macros" ]
strum_phf = [ "strum/std", "strum/phf", "strum/strum_macros" ]
# zzz : qqq : review features

derive_clone_dyn = [ "clone_dyn/enabled" ]
# derive_clone_dyn_use_std = [ "derive_clone_dyn", "clone_dyn/use_std" ]
# derive_clone_dyn_no_std = [ "derive_clone_dyn", "clone_dyn/no_std" ]
# derive_clone_dyn_use_alloc = [ "derive_clone_dyn", "clone_dyn/use_alloc" ]

derive_from = [ "derive_tools_meta/derive_from" ]
derive_index = [ "derive_tools_meta/derive_index" ]
derive_index_mut = [ "derive_tools_meta/derive_index_mut" ]
derive_inner_from = [ "derive_tools_meta/derive_inner_from" ]
derive_new = [ "derive_tools_meta/derive_new" ]

derive_not = [ "derive_tools_meta/derive_not" ]

derive_phantom = [ "derive_tools_meta/derive_phantom" ]

parse_display = [ "parse-display" ]


[dependencies]

## external
derive_more = { version = "~1.0.0-beta.6", optional = true, default-features = false, features = [ "debug" ] }
strum = { version = "~0.25", optional = true, default-features = false }
# strum_macros = { version = "~0.25.3", optional = true, default-features = false }
parse-display = { version = "~0.8.2", optional = true, default-features = false }


## internal
derive_tools_meta = { workspace = true, optional = true, features = [] }
variadic_from = { workspace = true, optional = true, features = [] }
clone_dyn = { workspace = true, optional = true, features = [ "clone_dyn_types", "clone_dyn_meta" ] }


[dev-dependencies]
test_tools = { workspace = true }

[build-dependencies]
cfg_aliases = "0.1.1"
