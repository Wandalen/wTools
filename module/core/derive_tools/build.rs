//! To avoid messing up with long logical expressions in the codebase.

use cfg_aliases::cfg_aliases;

fn main()
{
  // Setup cfg aliases
  cfg_aliases!
  {
    // Platforms
    // wasm : { target_arch = "wasm32" },
    // android : { target_os = "android" },
    // macos : { target_os = "macos" },
    // linux : { target_os = "linux" },
    all_derives:
    {
      all
      (
        feature = "derive_as_mut",
        feature = "derive_as_ref",
        feature = "derive_deref",
        feature = "derive_deref_mut",
        feature = "derive_from",
        feature = "derive_index",
        feature = "derive_index_mut",
        feature = "derive_inner_from",
        feature = "derive_variadic_from",
        feature = "derive_not",
        feature = "derive_reflect",
        feature = "derive_phantom"
      )
    },
    any_derive :
    {
      any
      (
        feature = "derive_as_mut",
        feature = "derive_as_ref",
        feature = "derive_deref",
        feature = "derive_deref_mut",
        feature = "derive_from",
        feature = "derive_index",
        feature = "derive_index_mut",
        feature = "derive_inner_from",
        feature = "derive_variadic_from",
        feature = "derive_not",
        feature = "derive_reflect",
        feature = "derive_phantom"
      )
    },
  }
}
