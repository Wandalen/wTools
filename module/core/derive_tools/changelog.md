# Changelog

### 2025-07-01
*   **Increment 4:** Performed final verification and addressed remaining issues in `derive_tools`.
    *   Resolved `#[display]` attribute parsing error by fixing attribute filtering in `derive_tools_meta/src/derive/from/field_attributes.rs` and `item_attributes.rs`.
    *   Resolved `From` trait bound error in `derive_tools_trivial.rs` example by adding `#[derive(From)]` to `Struct1`.
    *   Resolved "cannot find trait" errors by adding `pub use` statements for `VariadicFrom`, `InnerFrom`, `New`, `AsMut`, `AsRef`, `Deref`, `DerefMut`, `Index`, `IndexMut`, `Not`, `PhantomData` in `derive_tools/src/lib.rs`.
    *   Resolved `IndexMut` test issues by activating and correcting the `struct_named.rs` test (changing `#[index]` to `#[index_mut]`).
    *   Temporarily disabled the `PhantomData` derive macro and its doc comments in `derive_tools_meta/src/lib.rs` to resolve `E0392` and clippy warnings, as it requires a re-design.
    *   Created a `task.md` proposal for `module/core/clone_dyn` to address the `clippy::doc_markdown` warning in its `Readme.md`, as direct modification is out of scope.
    *   Confirmed `cargo test -p derive_tools` passes. `cargo clippy -p derive_tools` still fails due to the external `clone_dyn` issue.

*   [2025-07-01 11:13 UTC] Established baseline for derive_tools fix by commenting out `clone_dyn` tests and creating a task for `clone_dyn` test issues.

*   [2025-07-01 11:15 UTC] Added test matrices and purpose documentation for `AsMut` and `AsRef` derives.

*   [2025-07-01 11:18 UTC] Updated test command syntax in plan to correctly target internal test modules.

*   [2025-07-01 11:19 UTC] Re-enabled and fixed `as_mut` tests.

*   [2025-07-01 11:20 UTC] Updated test command syntax in plan to correctly target internal test modules.

*   [2025-07-01 11:21 UTC] Updated test command syntax in plan to correctly target internal test modules.

*   [2025-07-01 11:23 UTC] Updated test command syntax in plan to correctly target internal test modules.

*   [2025-07-01 11:24 UTC] Re-enabled and fixed `as_ref` tests.

*   [2025-07-01 11:25 UTC] Updated test command syntax in plan to correctly target internal test modules.

*   [2025-07-01 12:09 UTC] Added test matrices and purpose for Deref.

* [Increment 6 | 2025-07-01 13:25 UTC] Fixed `Deref` derive and tests for basic structs. Resolved `E0614`, `E0433`, `E0432` errors. Temporarily commented out `IsTransparentComplex` due to `E0207` (const generics issue in `macro_tools`). Isolated debugging with temporary test file was successful.

* [Increment 7 | 2025-07-01 13:45 UTC] Ensured `Deref` derive rejects enums with a compile-fail test. Removed enum-related test code and updated `deref.rs` macro to return `syn::Error` for enums. Fixed `Cargo.toml` dependency for `trybuild` tests.

* [Increment 8 | 2025-07-01 13:55 UTC] Marked `Deref` tests for generics and bounds as blocked due to `E0207` (unconstrained const parameter) in `macro_tools`. These tests remain commented out.
* [Increment 9 | 2025-07-01 13:58 UTC] Created and documented `DerefMut` test files (`basic_test.rs`, `basic_manual_test.rs`) with initial content and test matrices. Temporarily commented out `IsTransparentComplex` related code due to `E0207` (const generics issue in `macro_tools`).

* [Increment 10 | 2025-07-01 14:00 UTC] Fixed `DerefMut` derive and tests for basic structs. Resolved `E0277`, `E0614` errors. Ensured `DerefMut` derive rejects enums with a compile-fail test.
* [Increment 11 | 2025-07-01 14:05 UTC] Created and documented `From` test files (`basic_test.rs`, `basic_manual_test.rs`) with initial content and test matrices. Temporarily commented out `IsTransparentComplex` related code due to `E0207` (const generics issue in `macro_tools`).

* [Increment 11] Planned and documented `From` derive tests.

* [Increment 12] Implemented and fixed `From` derive macro.

* [Increment 13] Planned and documented `InnerFrom` and `New` tests.

* [Increment 14] Implemented and fixed `InnerFrom` derive macro.

* [Increment 15] Implemented and fixed `New` derive macro.

* [Increment 16] Planned and documented `Not`, `Index`, `IndexMut` tests.

* [Increment 17] Implemented and fixed `Not` derive macro.

* [Increment 18] Implemented and fixed `Index` and `IndexMut` derive macros.

* [Increment 19] Redesigned `PhantomData` derive macro to return an error when invoked, and added a compile-fail test to verify this behavior.

* [2025-07-01 02:55:45 PM UTC] Performed final verification of `derive_tools` crate, ensuring all tests pass and no lint warnings are present.

*   [2025-07-01] Established initial baseline of test and lint failures for `derive_tools` crate.

*   [2025-07-01] Fixed `macro_tools` `const` generics bug.

*   [Increment 7 | 2025-07-05 08:54 UTC] Re-enabled and fixed `IndexMut` derive macro, including `Index` trait implementation and `trybuild` tests.
