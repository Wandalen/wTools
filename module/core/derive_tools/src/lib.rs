#![ cfg_attr( feature = "no_std", no_std ) ]
#![ doc( html_logo_url = "https://raw.githubusercontent.com/Wandalen/wTools/master/asset/img/logo_v3_trans_square.png" ) ]
#![ doc( html_favicon_url = "https://raw.githubusercontent.com/Wandalen/wTools/alpha/asset/img/logo_v3_trans_square_icon_small_v2.ico" ) ]
#![ doc( html_root_url = "https://docs.rs/derive_tools/latest/derive_tools/" ) ]
#![ doc = include_str!( concat!( env!( "CARGO_MANIFEST_DIR" ), "/", "Readme.md" ) ) ]

// // xxx : implement derive new
//
// #[ derive( Debug, PartialEq, Default ) ]
// pub struct Property< Name >
// {
//   name : Name,
//   description : String,
//   code : isize,
// }
//
// /// generated by new
// impl< Name > Property< Name >
// {
//   #[ inline ]
//   pub fn new< Description, Code >( name : Name, description : Description, code : Code ) -> Self
//   where
//     Name : core::convert::Into< Name >,
//     Description : core::convert::Into< String >,
//     Code : core::convert::Into< isize >,
//   {
//     Self { name : name.into(), description : description.into(), code : code.into() }
//   }
// }

// #[ cfg( feature = "enabled" ) ]
// pub mod wtools;

#[ cfg( all( feature = "derive_more" ) ) ]
#[ allow( unused_imports ) ]
mod derive_more
{
  #[ cfg( feature = "derive_add" ) ]
  pub use ::derive_more::{ Add, Sub };
  #[ cfg( feature = "derive_add_assign" ) ]
  pub use ::derive_more::{ AddAssign, SubAssign };
  #[ cfg( feature = "derive_constructor" ) ]
  pub use ::derive_more::Constructor;
  #[ cfg( feature = "derive_error" ) ]
  pub use ::derive_more::Error;
  #[ cfg( feature = "derive_into" ) ]
  pub use ::derive_more::Into;
  // #[ cfg( feature = "derive_iterator" ) ]
  // pub use ::derive_more::Iterator;
  #[ cfg( feature = "derive_into_iterator" ) ]
  pub use ::derive_more::IntoIterator;
  #[ cfg( feature = "derive_mul" ) ]
  pub use ::derive_more::{ Mul, Div };
  #[ cfg( feature = "derive_mul_assign" ) ]
  pub use ::derive_more::{ MulAssign, DivAssign };
  #[ cfg( feature = "derive_sum" ) ]
  pub use ::derive_more::Sum;
  #[ cfg( feature = "derive_try_into" ) ]
  pub use ::derive_more::TryInto;
  #[ cfg( feature = "derive_is_variant" ) ]
  pub use ::derive_more::IsVariant;
  #[ cfg( feature = "derive_unwrap" ) ]
  pub use ::derive_more::Unwrap;

  // qqq : list all
  // qqq : make sure all features of derive_more is reexported
}

#[ doc( inline ) ]
#[ cfg( any( feature = "derive_variadic_from", feature = "type_variadic_from" ) ) ]
pub use variadic_from as variadic;

/// Namespace with dependencies.

#[ allow( unused_imports ) ]
#[ cfg( feature = "enabled" ) ]
pub mod dependency
{

  #[ doc( inline ) ]
  pub use ::derive_tools_meta;

  #[ doc( inline ) ]
  #[ cfg( feature = "derive_clone_dyn" ) ]
  pub use ::clone_dyn::{ self, dependency::* };

  #[ doc( inline ) ]
  #[ cfg( any( feature = "derive_variadic_from", feature = "type_variadic_from" ) ) ]
  pub use ::variadic_from::{ self, dependency::* };

  #[ doc( inline ) ]
  #[ cfg( feature = "derive_more" ) ]
  pub use ::derive_more;
  #[ doc( inline ) ]
  #[ cfg( feature = "derive_strum" ) ]
  pub use ::strum;
  #[ doc( inline ) ]
  #[ cfg( feature = "parse_display" ) ]
  pub use ::parse_display;

}

#[ doc( inline ) ]
#[ cfg( feature = "enabled" ) ]
#[ allow( unused_imports ) ]
pub use own::*;

/// Own namespace of the module.
#[ cfg( feature = "enabled" ) ]
#[ allow( unused_imports ) ]
pub mod own
{
  use super::*;
  #[ doc( inline ) ]
  pub use orphan::*;
  #[ cfg( feature = "derive_clone_dyn" ) ]
  #[ doc( inline ) ]
  pub use ::clone_dyn::orphan::*;
}

/// Orphan namespace of the module.
#[ cfg( feature = "enabled" ) ]
#[ allow( unused_imports ) ]
pub mod orphan
{
  use super::*;
  #[ doc( inline ) ]
  pub use exposed::*;
}

/// Exposed namespace of the module.
#[ cfg( feature = "enabled" ) ]
#[ allow( unused_imports ) ]
pub mod exposed
{
  use super::*;
  #[ doc( inline ) ]
  pub use prelude::*;

  #[ cfg( all( feature = "derive_more" ) ) ]
  #[ doc( inline ) ]
  pub use super::derive_more::*;

  #[ cfg( feature = "derive_strum" ) ]
  #[ doc( inline ) ]
  pub use ::strum::*;

  #[ cfg( any( feature = "derive_variadic_from", feature = "type_variadic_from" ) ) ]
  #[ doc( inline ) ]
  pub use ::variadic_from::exposed::*;

  #[ cfg( feature = "derive_strum" ) ]
  #[ doc( inline ) ]
  pub use ::strum::*;

  #[ cfg( feature = "derive_display" ) ]
  #[ doc( inline ) ]
  pub use ::parse_display::Display;

  #[ cfg( feature = "derive_from_str" ) ]
  #[ doc( inline ) ]
  pub use ::parse_display::FromStr;

  #[ cfg( feature = "derive_clone_dyn" ) ]
  #[ doc( inline ) ]
  pub use ::clone_dyn::exposed::*;

  #[ cfg( feature = "derive_clone_dyn" ) ]
  #[ doc( inline ) ]
  pub use ::clone_dyn;

  #[ doc( inline ) ]
  pub use ::derive_tools_meta::*;

  #[ doc( inline ) ]
  #[ cfg( feature = "derive_from" ) ]
  pub use ::derive_tools_meta::From;

  #[ doc( inline ) ]
  #[ cfg( feature = "derive_inner_from" ) ]
  pub use ::derive_tools_meta::InnerFrom;

}

/// Prelude to use essentials: `use my_module::prelude::*`.
#[ cfg( feature = "enabled" ) ]
#[ allow( unused_imports ) ]
pub mod prelude
{
  use super::*;
  #[ cfg( feature = "derive_clone_dyn" ) ]
  #[ doc( inline ) ]
  pub use ::clone_dyn;

  #[ cfg( feature = "derive_clone_dyn" ) ]
  #[ doc( inline ) ]
  pub use ::clone_dyn::prelude::*;

  #[ cfg( any( feature = "derive_variadic_from", feature = "type_variadic_from" ) ) ]
  #[ doc( inline ) ]
  pub use ::variadic_from::prelude::*;

}
