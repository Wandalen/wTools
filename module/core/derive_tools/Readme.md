# Module :: derive_tools

<!--{ generate.module_header{} }-->
<!--{ generate.module_header.start() }-->
 [![experimental](https://raster.shields.io/static/v1?label=&message=experimental&color=orange)](https://github.com/emersion/stability-badges#experimental) [![rust-status](https://github.com/Wandalen/wTools/actions/workflows/module_derive_tools_push.yml/badge.svg)](https://github.com/Wandalen/wTools/actions/workflows/module_derive_tools_push.yml) [![docs.rs](https://img.shields.io/docsrs/derive_tools?color=e3e8f0&logo=docs.rs)](https://docs.rs/derive_tools) [![Open in Gitpod](https://raster.shields.io/static/v1?label=try&message=online&color=eee&logo=gitpod&logoColor=eee)](https://gitpod.io/#RUN_PATH=.,SAMPLE_FILE=module/core/derive_tools/examples/derive_tools_trivial.rs,RUN_POSTFIX=--example%20/home/sakapoi/Документи/wTools_fork/module/core/derive_tools/examples/derive_tools_trivial/https://github.com/Wandalen/wTools) [![discord](https://img.shields.io/discord/872391416519737405?color=eee&logo=discord&logoColor=eee&label=ask)](https://discord.gg/m3YfbXpUUY)
<!--{ generate.module_header.end }-->

### Basic use-case

<!-- {{# generate.module{} #}} -->

```rust
# #[ cfg( all( feature = "derive_from", feature = "derive_inner_from", feature = "derive_display", feature = "derive_from_str" ) ) ]
{
  use derive_tools::*;

  #[ derive( From, InnerFrom, Display, FromStr, PartialEq, Debug ) ]
  #[ display( "{a}-{b}" ) ]
  struct Struct1
  {
    a : i32,
    b : i32,
  }

  // derived InnerFrom
  let src = Struct1 { a : 1, b : 3 };
  let got : ( i32, i32 ) = src.into();
  let exp = ( 1, 3 );
  assert_eq!( got, exp );

  // derived From
  let src : Struct1 = ( 1, 3 ).into();
  let got : ( i32, i32 ) = src.into();
  let exp = ( 1, 3 );
  assert_eq!( got, exp );

  // derived Display
  let src = Struct1 { a : 1, b : 3 };
  let got = format!( "{}", src );
  let exp = "1-3";
  println!( "{}", got );
  assert_eq!( got, exp );

  // derived FromStr
  use std::str::FromStr;
  let src = Struct1::from_str( "1-3" );
  let exp = Ok( Struct1 { a : 1, b : 3 } );
  assert_eq!( src, exp );

}
```

### To add to your project

```sh
cargo add derive_tools
```

### Try out from the repository

```sh
git clone https://github.com/Wandalen/wTools
cd wTools
cd examples/derive_tools_trivial
cargo run
```
