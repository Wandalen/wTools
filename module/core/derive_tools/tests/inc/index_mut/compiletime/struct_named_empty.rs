use derive_tools::IndexMut;

#[ derive( IndexMut ) ] 
struct EmptyStruct 
{    
}

fn main()
{  
}
