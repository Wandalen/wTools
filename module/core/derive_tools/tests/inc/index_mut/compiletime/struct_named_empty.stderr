error: proc-macro derive panicked
 --> tests/inc/index_mut/compiletime/struct_named_empty.rs:3:12
  |
3 | #[ derive( IndexMut ) ]
  |            ^^^^^^^^
  |
  = help: message: not implemented: IndexMut not implemented for Unit
